src/main/java/com/facebook/openwifirrm/modules/Modeler.java

```diff
@@ -299,8 +304,18 @@ private void processData(InputData data) {
 				if (state != null) {
 					try {
 						State stateModel = gson.fromJson(state, State.class);
-						dataModel.latestState
-							.put(record.serialNumber, stateModel);
+						List<State> latestStatesList = dataModel.latestStates
+							.computeIfAbsent(
+								record.serialNumber,
+								k -> Collections
+									.synchronizedList(new LinkedList<>())
+							);
+						if (latestStatesList.size() >= params.stateBufferSize) {
```

Comment:
this should be a `while` loop

Reply:
I guess `if` statement also works here? Because it is in a for loop 
```
for (KafkaRecord record : data.records)
```
at each iteration we only added one `stateModel` so only one `stateModel` is to be removed.

Reply:
It is possible that there are more elements in this list than should be there, the `while` loop will always truncate it to the correct size